File: src/plan.ts
/** 跨年便利纯函数（M4）：今年清单、漏访匹配、套名候选（D17/D18/F-9） */

import type { LatLng } from './geo.js';
import { haversineM } from './geo.js';
import type { HouseNode, SessionData } from './state.js';

export interface PlanItem {
  name: string; // 可为空，UI 显示"户N"兜底
  pos: LatLng | null; // null=手动添加（无坐标，不参与自动匹配，仅在漏访区提醒）
}

export interface Plan {
  year: number;
  items: PlanItem[];
  createdAt: number;
  updatedAt: number;
}

export const PLAN_MATCH_M = 10; // 与 D10 节点合并阈值一致

/** 从往年会话生成今年清单（F-1） */
export function planFromSession(
  prev: SessionData,
  year: number,
  now = Date.now(),
): Plan {
  return {
    year,
    createdAt: now,
    updatedAt: now,
    items: prev.nodes.map((n) => ({ name: n.name, pos: n.pos })),
  };
}

export interface MatchResult {
  visited: { item: PlanItem; nodeId: string }[];
  missing: PlanItem[]; // 疑似漏访
}

/** 贪心一对一匹配：item×node 距离升序，≤threshold 配对，每方最多一次（D18 回顾页对比） */
export function matchPlan(
  s: SessionData,
  plan: Plan,
  threshold = PLAN_MATCH_M,
): MatchResult {
  const cands: { i: number; j: number; d: number }[] = [];
  plan.items.forEach((item, i) => {
    const pos = item.pos;
    if (!pos) return; // 无坐标项不参与自动匹配
    s.nodes.forEach((n, j) => {
      cands.push({ i, j, d: haversineM(pos, n.pos) });
    });
  });
  cands.sort((a, b) => a.d - b.d);
  const usedItem = new Set<number>();
  const usedNode = new Set<number>();
  const visited: { item: PlanItem; nodeId: string }[] = [];
  for (const c of cands) {
    if (c.d > threshold) break;
    if (usedItem.has(c.i) || usedNode.has(c.j)) continue;
    usedItem.add(c.i);
    usedNode.add(c.j);
    visited.push({ item: plan.items[c.i], nodeId: s.nodes[c.j].id });
  }
  const missing = plan.items.filter((_, i) => !usedItem.has(i));
  return { visited, missing };
}

export interface NameCandidate {
  name: string;
  distM: number;
}

/** 套名候选：去年**有名字**的户按距离升序取前 top 个（D17） */
export function nameCandidates(
  nodePos: LatLng,
  prevNodes: HouseNode[],
  top = 3,
): NameCandidate[] {
  return prevNodes
    .filter((n) => n.name.trim() !== '')
    .map((n) => ({ name: n.name, distM: haversineM(nodePos, n.pos) }))
    .sort((a, b) => a.distM - b.distM)
    .slice(0, top);
}

File: android/core/src/main/kotlin/io/github/chenchen913/baibai/core/plan/Plan.kt
package io.github.chenchen913.baibai.core.plan

import io.github.chenchen913.baibai.core.geo.Geo
import io.github.chenchen913.baibai.core.model.HouseNode
import io.github.chenchen913.baibai.core.model.LatLng
import io.github.chenchen913.baibai.core.model.Plan
import io.github.chenchen913.baibai.core.model.PlanItem
import io.github.chenchen913.baibai.core.model.SessionData

/** 跨年便利纯函数（对应网页版 plan.ts：清单生成/漏访匹配/套名候选，D17/D18/F-9） */

const val PLAN_MATCH_M = 10.0 // 与 D10 节点合并阈值一致

data class VisitedMatch(val item: PlanItem, val nodeId: String)

data class MatchResult(val visited: List<VisitedMatch>, val missing: List<PlanItem>)

object PlanOps {

    /** 从往年会话生成今年清单（F-1） */
    fun planFromSession(prev: SessionData, year: Int, now: Long = System.currentTimeMillis()): Plan =
        Plan(
            year = year,
            items = prev.nodes.map { PlanItem(name = it.name, pos = it.pos) },
            createdAt = now,
            updatedAt = now,
        )

    /** 贪心一对一匹配：item×node 距离升序，≤threshold 配对，每方最多一次（D18 回顾页对比） */
    fun matchPlan(s: SessionData, plan: Plan, threshold: Double = PLAN_MATCH_M): MatchResult {
        data class Cand(val i: Int, val j: Int, val d: Double)
        val cands = mutableListOf<Cand>()
        plan.items.forEachIndexed { i, item ->
            val pos = item.pos ?: return@forEachIndexed // 无坐标项不参与自动匹配
            s.nodes.forEachIndexed { j, n ->
                cands.add(Cand(i, j, Geo.haversineM(pos, n.pos)))
            }
        }
        cands.sortBy { it.d }
        val usedItem = mutableSetOf<Int>()
        val usedNode = mutableSetOf<Int>()
        val visited = mutableListOf<VisitedMatch>()
        for (c in cands) {
            if (c.d > threshold) break
            if (c.i in usedItem || c.j in usedNode) continue
            usedItem.add(c.i)
            usedNode.add(c.j)
            visited.add(VisitedMatch(plan.items[c.i], s.nodes[c.j].id))
        }
        val missing = plan.items.filterIndexed { i, _ -> i !in usedItem }
        return MatchResult(visited, missing)
    }

    data class NameCandidate(val name: String, val distM: Double)

    /** 套名候选：去年**有名字**的户按距离升序取前 top 个（D17） */
    fun nameCandidates(nodePos: LatLng, prevNodes: List<HouseNode>, top: Int = 3): List<NameCandidate> =
        prevNodes
            .filter { it.name.isNotBlank() }
            .map { NameCandidate(it.name, Geo.haversineM(nodePos, it.pos)) }
            .sortedBy { it.distM }
            .take(top)
}
